packages/block-library/src/utils/use-template-context.js

```diff
@@ -0,0 +1,325 @@
+		for ( let i = 0; i < taxonomyPart.length; i++ ) {
+			if ( taxonomyPart[ i ] === '-' ) {
+				dashIndices.push( i );
+			}
+		}
+
+		if ( dashIndices.length > 0 ) {
+			// If there's only one dash, treat it as taxonomy.
+			// If there are multiple dashes, use the last one as the split point.
+			if ( dashIndices.length === 1 ) {
+				// Single dash
+				// e.g. taxonomy-product-category -> taxonomy: "product-category", term: null
+				return buildTemplateContext( { taxonomy: taxonomyPart } );
+			}
+
+			// Multiple dashes
+			// e.g. taxonomy-product-category-electronics -> taxonomy: "product-category", term: "electronics"
+			const lastDashIndex = dashIndices[ dashIndices.length - 1 ];
+			const taxonomy = taxonomyPart.substring( 0, lastDashIndex );
+			const termSlug = taxonomyPart.substring( lastDashIndex + 1 );
+
+			return buildTemplateContext( { taxonomy, termSlug } );
+		}
+
+		// No dashes, so the entire part is the taxonomy
+		return buildTemplateContext( { taxonomy: taxonomyPart } );
+	}
+
+	// Check for built-in taxonomy patterns
+	// e.g. category, tag
+	if ( templateSlug === 'category' ) {
+		return buildTemplateContext( { taxonomy: 'category' } );
+	}
+	if ( templateSlug === 'tag' ) {
+		return buildTemplateContext( { taxonomy: 'post_tag' } );
+	}
+
+	// Check for specific term patterns for built-in taxonomies
+	// e.g. category-news, tag-featured
+	if ( templateSlug.startsWith( 'category-' ) ) {
+		return buildTemplateContext( {
+			taxonomy: 'category',
+			termSlug: templateSlug.substring( 9 ),
+		} );
+	}
+	if ( templateSlug.startsWith( 'tag-' ) ) {
+		return buildTemplateContext( {
+			taxonomy: 'post_tag',
+			termSlug: templateSlug.substring( 4 ),
+		} );
+	}
+
+	return DEFAULT_CONTEXT;
+}
+
+/**
+ * Template slug parser that validates taxonomies exist.
+ *
+ * @param {string}   templateSlug The template slug to parse.
+ * @param {Function} getTaxonomy  Function to get taxonomy record.
+ * @return {Object} Object containing taxonomy, termSlug, isAuthor, and authorSlug.
+ */
+export function parseTemplateSlugWithValidation( templateSlug, getTaxonomy ) {
+	if ( ! templateSlug ) {
+		return DEFAULT_CONTEXT;
+	}
+
+	if ( templateSlug.startsWith( 'taxonomy-' ) ) {
+		const taxonomyPart = templateSlug.substring( 9 );
+		const dashIndices = [];
+
+		// Find all dash positions.
+		for ( let i = 0; i < taxonomyPart.length; i++ ) {
+			if ( taxonomyPart[ i ] === '-' ) {
+				dashIndices.push( i );
+			}
+		}
+
+		// If we have dashes, try different split points to find a valid taxonomy
+		if ( dashIndices.length > 0 ) {
+			for ( let i = 0; i < dashIndices.length; i++ ) {
+				const splitIndex = dashIndices[ i ];
+				const potentialTaxonomy = taxonomyPart.substring(
+					0,
+					splitIndex
+				);
+				const potentialTermSlug = taxonomyPart.substring(
+					splitIndex + 1
+				);
+
+				// Check if this taxonomy exists
+				const taxonomyRecord = getTaxonomy( potentialTaxonomy );
+				if ( taxonomyRecord ) {
```

Comment:
This fails when we have taxonomies like `taxonomy-one` and `taxonomy-one-more`. So if we're in a template for the second example it will match the first one because it also exists. It will fail later on though to fetch the terms etc.. as it will search for a term `more` as part of `taxonomy-one`.